src/RegistryCompilerPass.php

```diff
@@ -58,23 +73,35 @@ private function buildRegistry(ContainerBuilder $container, $registryId, array $
 					}
 				}
 
-				$collection->add($id, $options);
+				$collection->add([$id, $optionsResolver->getMetadata($options)], $options);
 			}
 
 		}
 
 		$targetId = $registryId;
 		$methodName = $tagOptions['method'];
+		$useMetadata = null !== $tagOptions['metadata'];
+
 		if ($tagOptions['use_collection']) {
 			$targetId = "{$registryId}.collection";
-			$container->setDefinition($targetId, new Definition(\ArrayObject::class))->setPublic(false);
+
+			$collectionClass = $useMetadata ? MetaDataIterator::class : \ArrayObject::class;
+			$container->setDefinition($targetId, new Definition($collectionClass))->setPublic(false);
```

Comment:
The only use case for `use_collection` I can think of is for services that require all the providers in the constructor — if you’d like to stick with the convention that required dependencies are put in the constructor, and optional ones are inserted via setters. 

But as I said, you can add your collection manually and tag it instead of the repository.